### PR TITLE
#159102247 Implement Login Authentication

### PR DESCRIPTION
**Pull request by ipheghe**
#### What Does This PR Do?
This PR implements login authentication with Andela email address

#### Description Of Task To Be Completed
1. Write code to authenticate a user on login action
2. Add actions for login function
3. Add user reducer

#### Any Background Context You Want To Provide?
Before now, a user was not properly authenticated before login into the application

#### How can this be manually tested?
- Git clone the repo
- Run scripts/start.sh command to start the application
- Click on the login button on the Login page

#### What are the relevant pivotal tracker stories?
[#159102247](https://www.pivotaltracker.com/story/show/159102247)

#### Screenshot(s)
N/A

### CODE REVIEW

File: client/pages/Dashboard/index.jsx [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+  /**
+   * React Lifecycle hook
+   *
+   * @param {nextProps, prevState}
+   * @memberof Dashboard
+   * @returns {null}
+   */
+  static getDerivedStateFromProps(nextProps, prevState) {
+    if (nextProps.activeUser
+      && nextProps.activeUser === prevState.activeUser) {
```
Comment (BolajiOlajide) [inline client/pages/Dashboard/index.jsx]: This logic won't work. A probable thing to do is check that the fields on one object are the same as the second

screen shot 2018-08-08 at 2 02 22 pm

File: client/pages/Dashboard/index.jsx [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+      return null;
```
Comment (BolajiOlajide) [inline client/pages/Dashboard/index.jsx]: Shouldn't be returning null also. Since you don't want anything to happen if this conditions are met, why not put the condition in which you want something to happen first.

File: client/pages/Dashboard/index.jsx [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    }
+  };
+
+  /**
+   * Method to display login error
+   *
+   * @memberof Dashboard
+   * @returns {null}
+   */
+  displayLoginError = () => this.props.displayLoginErrorMessage();
```
Comment (BolajiOlajide) [inline client/pages/Dashboard/index.jsx]: Why can't you call this method directly, since it's just referring to another method passed in?

File: client/pages/Dashboard/index.jsx [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    if (isTokenExpired() || !isLoggedIn()) {
+      return (
+        <Redirect to="/login" />
+      );
+    }
+
+    if (search === '?error=failed+to+create+user+token') {
+      this.displayLoginError();
```
Comment (BolajiOlajide) [inline client/pages/Dashboard/index.jsx]: If this block ever gets run, it'll still try to render the html because you didn't add a return statement here. Is that the desired effect?

File: client/pages/Dashboard/index.jsx [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    return (
+      <div className="wrapper">
+        <Header />
+          <Switch>
+            {this.redirectUser()}
```
Comment (BolajiOlajide) [inline client/pages/Dashboard/index.jsx]: Why is this here?

File: client/pages/Login/index.jsx [hunk 6]
```diff
@@ -1,14 +1,30 @@
...
 
 /**
  * Renders the Login component
  *
  * @returns {JSX} JSX
  * @memberof Login
  */
-const Login = () => (
+const Login = (props) => {
+  const redirectUrl = `${process.env.ANDELA_API_BASE_URL}/login?redirect_url=${process.env.BASE_URL}`;
```
Comment (BolajiOlajide) [inline client/pages/Login/index.jsx]: I think it's high time we have a global config object where we can all the variables we need. Similar to how it was on learning-map.

File: client/reducers/initialState.js [hunk 7]
```diff
@@ -1,9 +1,14 @@
 export default {
-  access: { isAuthenticated: false, user: {} },
+  access: {
```
Comment (BolajiOlajide) [inline client/reducers/initialState.js]: What's the difference between the user object in `access` and the `activeUser` object?

File: client/routes.js [hunk 8]
```diff
@@ -14,12 +13,10 @@ import NotFound from './components/common/NotFound';
...
-    <App>
-      <Route exact path="/login" component={Login} />
-
-      <Route component={NotFound} />
-    </App>
-  </Switch>
+    <Route exact path="/login" component={Login} />
+    <Route path="/" component={Dashboard} />
+    <Route path="*" component={NotFound} />
+</Switch>
```
Comment (BolajiOlajide) [inline client/routes.js]: This isn't indented properly. Check the indentation here

File: client/routes.js [hunk 9]
```diff
@@ -3,9 +3,8 @@ import React from 'react';
 import { Route, Switch } from 'react-router-dom';
 
 // components
-import App from './components/App';
```
Comment (BolajiOlajide) [inline client/routes.js]: I'm not sure how things are going to work but we need this App component to wrap majority of the components because w/o it no component can access the modal component @williamolojede created.

File: client/tests/cypress/integration/components/HomePage.spec.jsx [hunk 10]
```diff
@@ -1,15 +0,0 @@
-describe('Home Page', () => {
```
Comment (BolajiOlajide) [inline client/tests/cypress/integration/components/HomePage.spec.jsx]: Why did you delete this?

File: client/webpack.config.dev.js [hunk 11]
```diff
@@ -26,6 +26,8 @@ module.exports = {
           G_SUITE_DOMAIN: JSON.stringify(process.env.G_SUITE_DOMAIN),
           CLIENT_ID: JSON.stringify(process.env.CLIENT_ID),
           API_URI: JSON.stringify(process.env.REACT_APP_API_URI),
+          ANDELA_API_BASE_URL: JSON.stringify(process.env.ANDELA_API_BASE_URL),
```
Comment (BolajiOlajide) [inline client/webpack.config.dev.js]: You know what, on second thought, we can leave this here.

Comment (BolajiOlajide) [reply client/webpack.config.dev.js]: Though I feel some variables like the BASE URL aren't sensitive like that so they can live in a config object.

File: client/webpack.config.dev.js [hunk 12]
```diff
@@ -58,21 +60,24 @@ module.exports = {
...
+      },
       {
         test: /\.svg(\?v=\d+\.\d+\.\d+)?$/, loader: 'url?limit=10000&mimetype=image/svg+xml',
       },
     ],
   },
   devServer: {
     host: '0.0.0.0',
-    publicPath: 'http://0.0.0.0:9000/',
+    publicPath: 'http://0.0.0.0:3000/',
```
Comment (BolajiOlajide) [inline client/webpack.config.dev.js]: This is going to be a breaking change and everyone should be aware of this.

If possible, let this be reflected in the README.

File: .eslintrc.json [hunk 13]
```diff
@@ -5,18 +5,20 @@
...
+    "window": true
   },
   "rules": {
     "react/jsx-filename-extension": [1, { "extensions": [".js", ".jsx"] }],
     "object-curly-newline": ["error", {
-        "ObjectExpression": 0,
-        "ObjectPattern": { "multiline": true },
-        "ExportDeclaration": { "multiline": true, "minProperties": 3 }
+      "ObjectExpression": { "multiline": true, "minProperties": 2 },
+      "ObjectPattern": { "multiline": true, "minProperties": 2 },
```
[the remark this answers is not in the document]
Comment (ipheghe) [reply .eslintrc.json]: yea. That is the idea.

File: client/assets/style.scss [hunk 14]
```diff
@@ -56,82 +57,94 @@
...
   align-items: center;
-  color: #fff;
+  color: #FFF;
 }
 
 .page-header {
-    width: 100vw;
-    height: 300px;
+  width: 100vw;
+  height: 300px;
```
[the remark this answers is not in the document]
Comment (ipheghe) [reply client/assets/style.scss]: ok

File: client/webpack.config.dev.js (hunk 12 above)
Comment (ipheghe) [reply client/webpack.config.dev.js]: Was supposed to change that before I pushed, my port 9000 is being used by another application.

File: client/pages/Dashboard/index.jsx (hunk 5 above)
Comment (ipheghe) [reply client/pages/Dashboard/index.jsx]: `this.redirectUser` is a function that redirects an authenticated user to the default first page.

File: client/tests/cypress/integration/components/HomePage.spec.jsx (hunk 10 above)
Comment (ipheghe) [reply client/tests/cypress/integration/components/HomePage.spec.jsx]: It has the same test as `login.spec.js` file.

File: client/routes.js (hunk 9 above)
Comment (ipheghe) [reply client/routes.js]: I will sync with @williamolojede about this.

File: client/tests/cypress/integration/components/HomePage.spec.jsx [hunk 15]
```diff
@@ -1,15 +0,0 @@
...
-    cy.wait(1000);
-
-    // Should find the text "ANDELA" on the landing page
-    cy.contains('ANDELA');
-
-    // Should find the sign in button and click on it
-    cy.contains('SIGN IN').click();
-  });
-});
-
```
[the remark this answers is not in the document]
Comment (ipheghe) [reply client/tests/cypress/integration/components/HomePage.spec.jsx]: ok. I will do that

File: client/routes.js (hunk 9 above)
Comment (ipheghe) [reply client/routes.js]: Yes, It has been clarified.
